Template and store replicas count as service configuration in Metalk8s cluster

Ebaneck opened the pull request:
**Component**:

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->
'salt', 'charts', 

**Context**: 

See #2255 

**Summary**:

Template and render the charts to consume service configuration values directly.

**Acceptance criteria**: 


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2255 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

Ebaneck committed:
salt: deploy service configurations using configmaps in a metalk8s cluster

Initially, we do not deploy and manage service configurations using
configmaps.

This commit adds generic configmaps with service related configurations
such that the end user can configure cluster settings directly by editing
the related configmaps

Ebaneck committed:
salt: add a salt module to query configmap

This commit adds a salt module used to query a configmap object
within a metalk8s cluster

Once called with a service config parameter, this module returns
the value of the configuration

Ebaneck committed:
chart: Fix the render.py script to use argparse

This commit adds argparse to our render script and introduces
a new argument `--service-config` which can be called several times
accepting only two arguments(service name & service configmap name)

Ebaneck committed:
chart: template and render charts to use service values

The prometheus-operator chart is rendered using the following command:

./charts/render.py prometheus-operator --namespace metalk8s-monitoring charts/prometheus-operator.yaml --service-config grafana
metalk8s-grafana-config --service-config prometheus metalk8s-prometheus-config --service-config alertmanager metalk8s-alertmanager-config
charts/prometheus-operator/ > salt/metalk8s/addons/prometheus-operator/deployed/chart.sls

Service configuration values are passed using `--service-config` argument.
When rendering this chart, we must pass the following services(grafana, prometheus, alertmanager)
including the configmap that holds their respective service & cluster configurations

Closes: #2255

bert-e commented:
# Hello ebaneck,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2255-template-and-store-replicas-as-service-configuration` into
`development/2.5`
* [w/2.6/feature/2255-template-and-store-replicas-as-service-configuration](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/feature/2255-template-and-store-replicas-as-service-configuration) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

gdemonet commented on charts/render.py:
```diff
@@ -114,25 +120,45 @@ def keep_doc(doc):
+        help="Ex: --service-config service_name service_configmap_name"
+    )
+    parser.add_argument('path')
+    args, extra = parser.parse_known_args()
```
Not sure you want to allow (and ignore) extra arguments. Just using `args = parser.parse_args()` will fail if unknown arguments are provided.

Ebaneck replied:
Well, we redirect the output to a file using something like `> salt/metalk8s/addons/prometheus-operator/deployed/chart.sls`

We actually want to omit this extra argument

TeddyAndrieux replied:
The redirect is catch by your shell directly and will **not** be part of the arguments python will get

Ebaneck replied:
ok

gdemonet commented on charts/render.py:
```diff
@@ -114,25 +120,45 @@ def keep_doc(doc):
-    (name, namespace, values, path) = sys.argv[1:]
+    parser = argparse.ArgumentParser()
+    parser.add_argument('name', help="Denotes the name of the chart")
+    parser.add_argument('namespace', help="Metalk8s namespace to deploy unto")
```
Let's make this optional:

```suggestion
    parser.add_argument(
        '-n',
        '--namespace',
        default="default",
        help="Namespace to deploy this chart in",
    )
```

NicolasT replied:
```suggestion
    parser.add_argument('namespace', help="MetalK8s namespace to deploy into")
```
Also, this is not strictly a `MetalK8s` namespace.

slaperche-scality commented on charts/render.py:
```diff
@@ -27,17 +28,22 @@
 
 
-BOILERPLATE = '''
+BOILERPLATE = """
```
Even though I personally prefer `"""` over `'''`, let's be consistent here and use `'''` (since  `SUB_BOILERPLATE` and `SUB_BOILERPLATE_END` already use `'''`)

gdemonet commented on charts/render.py:
```diff
@@ -114,25 +120,45 @@ def keep_doc(doc):
+        dest="service_config",
+        help="Ex: --service-config service_name service_configmap_name"
+    )
+    parser.add_argument('path')
```
Let's add some help as well:

```suggestion
    parser.add_argument('path', help="Path to the chart directory")
```

gdemonet commented on charts/render.py:
```diff
@@ -114,25 +120,45 @@ def keep_doc(doc):
+        nargs=2,
+        dest="service_config",
+        help="Ex: --service-config service_name service_configmap_name"
+    )
```
Cool stuff! Could add a small explanation about what those "service-config"s are in the help message (and adjust the example to make it understandable that `service_name` is actually the Jinja variable which will hold this ConfigMap's contents).

(minor) To make it even friendlier to use/read in code, you can cast parsed arguments into arbitrary types:

```python
from collections import namedtuple

ServiceConfig = namedtuple('ServiceConfig', ('var_name', 'configmap_name')

# When building the parser...
parser.add_argument(
    '--service-config',
    action='append',
    nargs=2,
    type=ServiceConfig,
    dest="service_configs",
    help="...",
)

# When printing the Jinja declarations...
sys.stdout.write(
    BOILERPLATE.format(
        configlines='\n'.join(
            ("{{%- set {s.var_name} = salt.metalk8s_service_configurator"
             ".get_service_conf('{namespace}', '{s.configmap_name}') %}}").format(
                 s=service_config, namespace=args.namespace
            ) for service_config in args.service_configs
        )
    ).lstrip()
)
```

Note that you don't need to inject curly braces with `format` arguments, simply double them (`{{` or `}}`) to escape the characters. (cf. https://docs.python.org/library/string.html#formatstrings)

gdemonet commented on charts/render.py:
```diff
@@ -114,25 +120,45 @@ def keep_doc(doc):
+                ) for p in args.service_config)
+        ).lstrip()
+    )
+    sys.stdout.write(SUB_BOILERPLATE)
```
Not sure the split between `BOILERPLATE` and `SUB_BOILERPLATE` is needed or helpful. Actually, since we now template in some variable declarations, it's not only boilerplate anymore. Maybe just call them `[START|END]_TEMPLATE` or `[START|END]_BLOCK`?

NicolasT replied:
I'm quite confused by this naming as well.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

TeddyAndrieux commented on salt/_modules/metalk8s_service_configurator.py:
```diff
@@ -0,0 +1,80 @@
+        salt-call metalk8s_service_configurator.service_conf "metalk8s-monitoring" "metalk8s-prometheus-config"
+    """
+
+    if configmap_name is None:
```
```suggestion
    if not configmap_name:
```
If user give `0` or `''` as name you want to raise the  same message

TeddyAndrieux commented on salt/_modules/metalk8s_service_configurator.py:
```diff
@@ -0,0 +1,80 @@
+        )
+    if not config:
+        raise CommandExecutionError(
+            'Expected configmap as yaml but got {}'.format(config)
```
I got what you mean but I think the message is not really clear maybe something like

```suggestion
            'Expected `config.yaml` as yaml in the ConfigMap {} but got {}'.format(configmap_name, config)
```

TeddyAndrieux commented on buildchain/buildchain/salt_tree.py:
```diff
@@ -221,6 +221,10 @@ def _get_parts(self) -> Iterator[str]:
 
+    Path('salt/metalk8s/addons/addons-service-config/configmap.sls'),
+    Path('salt/metalk8s/addons/addons-service-config/deployed.sls'),
+    Path('salt/metalk8s/addons/addons-service-config/init.sls'),
```
You should move these files, we said that state that should only run on the salt-master need to be called `deployed`.
Here you can call `addons.addons-service-config` and `addons.addons-service-config.configmap` and both will deploy kubernetes objects.

So here you should have
```
salt/metalk8s/addons/addons-service-config/deployed/init.sls
salt/metalk8s/addons/addons-service-config/deployed/configmap.sls
```
And that it, then you will be able to do `state.sls metalk8s.addons.addons-service-config.deployed`

Also IMO the `addons` in `addons-service-config` is a bit redundant

NicolasT replied:
Can we have these *under the applicable addons or whatnot directories* rather than somewhere completely else? This is not an 'addon'.

TeddyAndrieux commented on salt/_modules/metalk8s_service_configurator.py:
```diff
@@ -0,0 +1,80 @@
+
+import yaml
+from yaml.dumper import SafeDumper
+from yaml.representer import SafeRepresenter
```
Look like these are not really used here

TeddyAndrieux commented on salt/_modules/metalk8s_service_configurator.py:
```diff
@@ -0,0 +1,80 @@
+'''Metalk8s cluster and service configuration utility.'''
+
+import logging
+import re
```
Same, not used here

TeddyAndrieux commented on salt/_modules/metalk8s_service_configurator.py:
```diff
@@ -0,0 +1,80 @@
+import logging
+import re
+
+from salt.exceptions import CommandExecutionError
```
Please respect pylint import suggestion
```
C:  8, 0: third party import "import yaml" should be placed before "from salt.exceptions import CommandExecutionError" (wrong-import-order)
```

slaperche-scality replied:
Maybe we should setup pylint for Salt tree (or `_modules` at least)?
That would automatically flag unused import and import ordering.

That way we can enforce it in CI and focus our brain's bandwidth on functional issues.

TeddyAndrieux replied:
Agree we should but we need some extra stuff about "salt" stuffs like all the `__opts__` `__salt__` `__pillar__` ... but anyway agree we should have salt module checking in the CI so linting + unit test

gdemonet replied:
If this is something we really want to enforce, it would make sense to add it in CI. Especially since not everyone here uses `pylint` + the special Salt plugin to lint their Salt modules (I don't).

slaperche-scality replied:
I thin there is a pylint plugin for Salt, otherwise we can simply ignore Salt stuff by tweaking Pylint conf (you can probably consider some stuff as blackbox).

TeddyAndrieux replied:
https://github.com/saltstack/salt-pylint (Personnaly never tested)

slaperche-scality replied:
I think @jbertran tried it (but I may remember wrong)

NicolasT commented on buildchain/buildchain/salt_tree.py:
```diff
@@ -552,6 +556,7 @@ def _get_parts(self) -> Iterator[str]:
     Path('salt/_modules/metalk8s_package_manager_yum.py'),
     Path('salt/_modules/metalk8s_package_manager_apt.py'),
     Path('salt/_modules/metalk8s_volumes.py'),
+    Path('salt/_modules/metalk8s_service_configurator.py'),
```
Is it a `configurator` or is it about `service_configuration`?

Ebaneck replied:
service_configuration

NicolasT commented on salt/_modules/metalk8s_service_configurator.py:
```diff
@@ -0,0 +1,80 @@
+
+    if configmap_name is None:
+        raise Exception(
+            'Expected a configmap name but got {}'.format(configmap_name)
```
```suggestion
            'Expected a ConfigMap name but got {}'.format(configmap_name)
```
Similar changes below. Be consistent.

NicolasT commented on salt/_modules/metalk8s_service_configurator.py:
```diff
@@ -0,0 +1,80 @@
+    """Reads  a ConfigMap from the name specified
+
+    Arguments:
+        configmap_name: Configmap name
```
```suggestion
        configmap_name: ConfigMap name
```

NicolasT commented on salt/_modules/metalk8s_service_configurator.py:
```diff
@@ -0,0 +1,80 @@
+
+    Arguments:
+        configmap_name: Configmap name
+        namespace: the namespace where the configmap is found
```
```suggestion
        namespace: the Namespace where the ConfigMap is stored
```

NicolasT commented on salt/_modules/metalk8s_service_configurator.py:
```diff
@@ -0,0 +1,80 @@
+        )
+    except ValueError as exc:
+        raise CommandExecutionError(
+            'Failed to read Configmap object {} with error {}'
```
```suggestion
            'Failed to read ConfigMap object {}: {}'
```

NicolasT commented on salt/_modules/metalk8s_service_configurator.py:
```diff
@@ -0,0 +1,80 @@
+    except ValueError as exc:
+        raise CommandExecutionError(
+            'Failed to read Configmap object {} with error {}'
+            .format(configmap_name, str(exc))
```
Use a modifier in the template instead of `str` here.

gdemonet replied:
(for those who don't know what this means:

```python
'Failed to read ConfigMap object {}: {!s}'.format(configmap_name, exc)
```
)

NicolasT commented on salt/_modules/metalk8s_service_configurator.py:
```diff
@@ -0,0 +1,80 @@
+
+    if manifest is None:
+        raise CommandExecutionError(
+            'Expected Configmap object but got {}'.format(manifest)
```
```suggestion
            'Expected ConfigMap object but got {}'.format(manifest)
```

NicolasT commented on salt/_modules/metalk8s_service_configurator.py:
```diff
@@ -0,0 +1,80 @@
+        # we are only interested in the data section of the configmap for now
+        conf_section = manifest.get('data', {}).get('config.yaml', {})
+        config = yaml.safe_load(conf_section) or {}
+    except Exception as exc:
```
Do we really need to catch `Exception`?

Ebaneck replied:
True! no need to catch it

gdemonet replied:
You do need to catch some exception, in case loading of the ConfigMap as YAML fails. Nicolas' point was: don't use bare `Exception`, look for something more specific (like `yaml.YAMLError`).

Leaving something like:
```python
except Exception:
    pass
```

is worse than not doing anything (otherwise your code will break in very strange ways).

NicolasT commented on salt/_modules/metalk8s_service_configurator.py:
```diff
@@ -0,0 +1,80 @@
+        config = yaml.safe_load(conf_section) or {}
+    except Exception as exc:
+        raise CommandExecutionError(
+            'Failed to get Config section {} as manifest with error {}'
```
```suggestion
            'Failed to get config section {}: {}'
```

NicolasT commented on salt/_modules/metalk8s_service_configurator.py:
```diff
@@ -0,0 +1,80 @@
+    except Exception as exc:
+        raise CommandExecutionError(
+            'Failed to get Config section {} as manifest with error {}'
+            .format(configmap_name, str(exc))
```
Same as above, use a modifier in the template, no `str`.

NicolasT commented on salt/_modules/metalk8s_service_configurator.py:
```diff
@@ -0,0 +1,80 @@
+        raise CommandExecutionError(
+            'Expected configmap as yaml but got {}'.format(config)
+        )
+    return config  # a classic python dict
```
Too early. Could as well do some super basic validation of this value. Even more, we may want to pass the `apiVersion` and `kind` of the object we expect (at call site) and assert these are 'correct'. And over time, do proper validation of the whole structure.

NicolasT commented on salt/metalk8s/addons/addons-service-config/configmap.sls:
```diff
@@ -0,0 +1,41 @@
+#! metalk8s_kubernetes
```
As mentioned before, these shouldn't go here.

NicolasT commented on salt/metalk8s/addons/addons-service-config/configmap.sls:
```diff
@@ -0,0 +1,41 @@
+  name: metalk8s-grafana-config
+data:
+  config.yaml: |-
+    apiVersion: metalk8s.scality.com/v1alpha1
```
We may want to give these their own group. WDYT @gdemonet

gdemonet replied:
`addons.metalk8s.scality.com`? Or even more specialized, like `<addon-name>.addons.metalk8s.scality.com`?

NicolasT commented on salt/metalk8s/addons/addons-service-config/deployed.sls:
```diff
@@ -0,0 +1,5 @@
+# Create the namespace before deploying configmap objects
```
This is exactly why the default configs belong with the addons and not here.

Ebaneck replied:
ok

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @gdemonet

gdemonet committed:
charts: bypass unqouted Jinja dump to YAML

This hack is required to bypass the fact that dumping Jinja
to YAML does not work with the render script because of the presence
of unquote special characters e.g `{%`.

If we try to escape thess special characters, we end up with type(str)
always and in some special conditions, we need for example replicas count
to be of type(int)

bert-e commented:
# History mismatch



Merge commit #cf344d4b9804a93461b48ebb05b47f939c0f4f55 on the integration branch
`w/2.6/feature/2255-template-and-store-replicas-as-service-configuration` is merging a branch which is neither the current
branch `feature/2255-template-and-store-replicas-as-service-configuration` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `feature/2255-template-and-store-replicas-as-service-configuration` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented on salt/metalk8s/addons/prometheus-operator/deployed/service-configuration.sls:
```diff
@@ -0,0 +1,41 @@
+    kind: AlertmanagerConfig
+    spec:
+      deployment:
+        replicas: 1
```
IMO we should do some jinja here to check that these ConfigMap not already exists to not override customers changes if this state run (and then when we have this, this state could be included in the `init.sls` of each addons)

Ebaneck replied:
Done!

TeddyAndrieux commented on salt/metalk8s/orchestrate/downgrade/init.sls:
```diff
@@ -133,6 +133,13 @@ Sync module on salt-master:
     - name: saltutil.sync_all
     - saltenv: metalk8s-{{ dest_version }}
 
+Deploy Kubernetes service config objects:
+  salt.runner:
+  - name: state.orchestrate
+  - mods:
+    - metalk8s.service-configuration.deployed
+  - saltenv: metalk8s-{{ dest_version }}
```
You need to add some checks here because you don't want to do this if `dest_version < 2.5.0`

Ebaneck replied:
Done

TeddyAndrieux commented on salt/metalk8s/service-configuration/deployed/init.sls:
```diff
@@ -0,0 +1,10 @@
+
+# Create the Namespace before deploying ConfiMap objects
+# If we include service configurations in their respective addons deploy states,
+# then salt will fail while rendering the jinja since the Namespace object does
```
I think you should say that is that '"ConfigMap' does not exist yet" even if the first error you get is because the namespace do not exist the issue here is that the CM do not exist (even if you create the namespace you will get an error because the ConfigMap do not exist)

TeddyAndrieux commented on salt/metalk8s/orchestrate/downgrade/init.sls:
```diff
@@ -133,6 +133,23 @@ Sync module on salt-master:
+    persisted throughout. Now in `2.5.0` we have added ConfigMaps to store
+    service configurations. When running a downgrade to a version inferior to
+    `2.5.0` we do not need to deploy these ConfigMap objects. #}
+{#- Only run this step if `dest_version < 2.5.0` #}
```
No, we want to run it only if `dest_version >= 2.5.0` :)

Ebaneck replied:
Thinking this still works because we would have already deployed the ConfigMaps in the Bootstrap orchestrate before the downgrade.

TeddyAndrieux commented on salt/metalk8s/orchestrate/downgrade/init.sls:
```diff
@@ -133,6 +133,23 @@ Sync module on salt-master:
+    `2.5.0` we do not need to deploy these ConfigMap objects. #}
+{#- Only run this step if `dest_version < 2.5.0` #}
+
+{%- if salt.pkg.version_cmp(dest_version, '2.5.0') == -1  %}
```
```suggestion
{%- if salt.pkg.version_cmp(dest_version, '2.5.0') >= 0  %}
```

TeddyAndrieux commented on salt/_modules/metalk8s_service_configuration.py:
```diff
@@ -0,0 +1,105 @@
+        config = yaml.safe_load(conf_section) or {}
+    except yaml.YAMLError as exc:
+        raise CommandExecutionError(
+            'Invalid YAML format in Configmap {}: {!s}'.format(
```
```suggestion
            'Invalid YAML format in ConfigMap {}: {!s}'.format(
```

Ebaneck replied:
This is getting out of hand.. Going to fix

TeddyAndrieux commented on salt/_modules/metalk8s_service_configuration.py:
```diff
@@ -0,0 +1,105 @@
+        )
+    except Exception as exc:
+        raise CommandExecutionError(
+            'Failed loading `config.yaml` from Configmap {}: {!s}'.format(
```
```suggestion
            'Failed loading `config.yaml` from ConfigMap {}: {!s}'.format(
```

TeddyAndrieux commented on salt/_modules/metalk8s_service_configuration.py:
```diff
@@ -0,0 +1,105 @@
+            'Expected value {} for key kind, got {}'.format(
+                config['kind'], kind
+            )
+        )
```
I don't really get the logic here, because you return config if `kind` and `apiVersion` are both None, but if only one is None that will fail.

If I well understand what you want to do is if `kind` or `apiVersion` is `None` just dont check it 
If yes then you need to also check for `None` in the 2 last `if` because one of `apiVersion` or `kind` could be `None`

```suggestion
    if not apiVersion and not kind:
        return config

    if apiVersion and config['apiVersion'] != apiVersion:
        raise CommandExecutionError(
            'Expected value {} for key apiVersion, got {}'.format(
                apiVersion, config['apiVersion']
            )
        )

    if kind and config['kind'] != kind:
        raise CommandExecutionError(
            'Expected value {} for key kind, got {}'.format(
                kind, config['kind']
            )
        )
```

Also, it's only my opinion but for me it do not hurt to check for empty string or 0 ... for `apiVersion` and `kind` not only for `None`

Ebaneck replied:
Accepted and will fix after discussing it

TeddyAndrieux reviewed (commented):
Few minor comment but LGTM

TeddyAndrieux commented on salt/_modules/metalk8s_service_configuration.py:
```diff
@@ -0,0 +1,104 @@
+    if apiVersion and config['apiVersion'] != apiVersion:
+        raise CommandExecutionError(
+            'Expected value {} for key apiVersion, got {}'.format(
+                config['apiVersion'], apiVersion
```
```suggestion
                apiVersion, config['apiVersion']
```
Like I told you on slack, you expect the `apiVersion` args from this function, not the one from configmap

TeddyAndrieux commented on salt/_modules/metalk8s_service_configuration.py:
```diff
@@ -0,0 +1,104 @@
+    if kind and config['kind'] != kind:
+        raise CommandExecutionError(
+            'Expected value {} for key kind, got {}'.format(
+                config['kind'], kind
```
```suggestion
                kind, config['kind']
```

TeddyAndrieux commented on salt/metalk8s/orchestrate/downgrade/init.sls:
```diff
@@ -133,6 +133,23 @@ Sync module on salt-master:
+  - name: state.orchestrate
+  - mods:
+    - metalk8s.service-configuration.deployed
+  - saltenv: metalk8s-{{ dest_version }}
```
Just add a require_in on Deploy kubernetes objects here (you don't want to run the kubernetes object deployment if this state failed)

And also a require on Sync module on salt-master

TeddyAndrieux commented on salt/metalk8s/addons/prometheus-operator/deployed/service-configuration.sls:
```diff
@@ -0,0 +1,83 @@
+            spec:
+              deployment:
+                replicas: 1
+{%- endif %}
```
An sls cannot be empty or you will get an error (or maybe just a warning don't remember) and here if all config already exists after jinja rendering you will get empty sls.

So suggest to just add an `else` block with `test.succeed_without_changes` explaining that the configmap already exists nothing need to be done

TeddyAndrieux commented on salt/metalk8s/orchestrate/upgrade/init.sls:
```diff
@@ -115,6 +115,13 @@ Sync module on salt-master:
     - name: saltutil.sync_all
     - saltenv: metalk8s-{{ dest_version }}
 
+Deploy Kubernetes service config objects:
+  salt.runner:
+  - name: state.orchestrate
+  - mods:
+    - metalk8s.service-configuration.deployed
+  - saltenv: metalk8s-{{ dest_version }}
```
Require Sync module on salt-master here

TeddyAndrieux reviewed (approved):
LGTM

Ebaneck commented:
/approve

bert-e commented:
# History mismatch



Merge commit #cf344d4b9804a93461b48ebb05b47f939c0f4f55 on the integration branch
`w/2.6/feature/2255-template-and-store-replicas-as-service-configuration` is merging a branch which is neither the current
branch `feature/2255-template-and-store-replicas-as-service-configuration` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `feature/2255-template-and-store-replicas-as-service-configuration` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

Ebaneck commented:
/reset

Ebaneck commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2255-template-and-store-replicas-as-service-configuration` into
`development/2.5`
* [w/2.6/feature/2255-template-and-store-replicas-as-service-configuration](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/feature/2255-template-and-store-replicas-as-service-configuration) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @gdemonet







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @gdemonet







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet







*The following options are set:* **approve**

Ebaneck commented:
/status

bert-e commented:
# Status





*Status report is not available.*






*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



Please check the status of the associated issue None.

Goodbye ebaneck.
